src/GraphQL/Types/Composite/ComplexGraphType.cs

```diff
@@ -561,15 +561,30 @@ public virtual FieldType AddField(FieldType fieldType)
     /// <param name="type">The graph type of the field; if <see langword="null"/> then will be inferred from the specified expression via registered schema mappings.</param>
     [Obsolete("Please use another overload that receives only one of the 'nullable' or 'type' arguments. This method will be removed in v9.")]
     public virtual FieldBuilder<TSourceType, TProperty> Field<TProperty>(
-    string name,
-    Expression<Func<TSourceType, TProperty>> expression,
+        string name,
+        Expression<Func<TSourceType, TProperty>> expression,
         bool nullable,
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type? type) =>
+        Field(name, expression, (bool?)nullable, type);
+
+    private FieldBuilder<TSourceType, TProperty> Field<TProperty>(
+        string name,
+        Expression<Func<TSourceType, TProperty>> expression,
+        bool? nullable,
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type? type)
     {
         try
         {
-            if (type == null)
-                type = typeof(TProperty).GetGraphTypeFromType(nullable, this is IInputObjectGraphType ? TypeMappingMode.InputType : TypeMappingMode.OutputType);
+            if (type == null && nullable == null && GlobalSwitches.InferFieldNullabilityFromNRTAnnotations)
+            {
+                var typeInfo = AutoRegisteringHelper.GetTypeInformation(((MemberExpression)expression.Body).Member, this is IInputObjectGraphType);
+                type = typeInfo.ConstructGraphType();
+            }
```

Comment:
I think we need a little more code in here, along these lines -- see example below: (unfinished)

```suggestion
            if (type == null && nullable == null && GlobalSwitches.InferFieldNullabilityFromNRTAnnotations)
            {
                if (expression.Body is MemberExpression memberExpression)
                {
                    var typeInfo = AutoRegisteringHelper.GetTypeInformation(memberExpression.Member, this is IInputObjectGraphType);
                    type = typeInfo.ConstructGraphType();
                }
                else if (typeof(TProperty).IsValueType)
                {
                    nullable = Nullable.GetUnderlyingType(typeof(TProperty)) != null;
                    // call GetGraphTypeFromType
                }
                else // non-member expression such as: x => "Sample"
                {
                    nullable = false; // fallback to legacy behavior
                    // call GetGraphTypeFromType
                }
            }
```